sched/event/event_post.c

```diff
@@ -0,0 +1,154 @@
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_sem_post
+ ****************************************************************************/
+
+static inline_function int nxevent_sem_post(FAR sem_t *sem)
+{
+  int semcount;
+
+  nxsem_get_value(sem, &semcount);
+  if (semcount < 1)
+    {
+      return nxsem_post(sem);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_post
+ *
+ * Description:
+ *   Post one or more events to an event object.
+ *
+ *   This routine posts one or more events to an event object. All tasks
+ *   waiting on the event object event whose waiting conditions become
+ *   met by this posting immediately unpend.
+ *
+ *   Posting differs from setting in that posted events are merged together
+ *   with the current set of events tracked by the event object.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to post to event
+ *          - Set events to 0 will be considered as any,
+ *            waking up the waiting thread immediately.
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   It follows the NuttX internal error return policy:  Zero (OK) is
+ *   returned on success.  A negated errno value is returned on failure.
+ *
+ * Assumptions:
+ *   This function may be called from an interrupt handler.
+ *
+ ****************************************************************************/
+
+int nxevent_post(FAR nxevent_t *event, nxevent_mask_t events,
+                 nxevent_flags_t eflags)
+{
+  FAR nxevent_wait_t *wait;
+  FAR nxevent_wait_t *tmp;
+  irqstate_t flags;
+  bool bwaitany;
+  bool bpostall;
+  int ret = 0;
+
+  if (event == NULL)
+    {
+      return -EINVAL;
+    }
+
+  if (events == 0)
+    {
+      events = ~0;
+    }
+
+  flags = enter_critical_section();
+
+  event->events |= events;
+
+  if (!list_is_empty(&event->list))
+    {
+      bpostall = ((eflags & NXEVENT_POST_ALL) != 0);
+
+      /* Hold schedule lock here to avoid context switch if post high
+       * priority task.
+       */
+
+      sched_lock();
+
+      list_for_every_entry_safe(&event->list, wait, tmp,
+                                nxevent_wait_t, node)
+        {
+          bwaitany = ((wait->eflags & NXEVENT_WAIT_ALL) == 0);
+
+          if ((bwaitany && ((wait->expect & event->events) != 0))
+              || (!bwaitany &&
+                  ((wait->expect & event->events) == wait->expect)))
+            {
+              list_delete(&wait->node);
+
+              ret = nxevent_sem_post(&wait->sem);
```

Comment:
```suggestion
              ret = nxevent_sem_post(&wait->sem);
              if (ret < 0)
                {
                  continue;
                }
                
              list_delete(&wait->node);
```

Reply:
I think we should remove the `invalid` waiter from wait list.